public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx

```diff
@@ -1,45 +1,107 @@
 // Libraries
-import React, { memo } from 'react';
+import React, { PureComponent } from 'react';
 
 // Types
 import { AbsoluteTimeRange, QueryEditorProps } from '@grafana/data';
+import { InlineFormLabel } from '@grafana/ui';
 import { LokiDatasource } from '../datasource';
 import { LokiQuery } from '../types';
 import { LokiQueryField } from './LokiQueryField';
 
 type Props = QueryEditorProps<LokiDatasource, LokiQuery>;
 
-export const LokiQueryEditor = memo(function LokiQueryEditor(props: Props) {
-  const { query, data, datasource, onChange, onRunQuery } = props;
+interface State {
+  legendFormat: string;
+}
 
-  let absolute: AbsoluteTimeRange;
+export class LokiQueryEditor extends PureComponent<Props, State> {
+  // Query target to be modified and used for queries
+  query: LokiQuery;
 
-  if (data && data.request) {
-    const { range } = data.request;
-    absolute = {
-      from: range.from.valueOf(),
-      to: range.to.valueOf(),
-    };
-  } else {
-    absolute = {
-      from: Date.now() - 10000,
-      to: Date.now(),
+  absolute: AbsoluteTimeRange;
+
+  constructor(props: Props) {
+    super(props);
+    // Use default query to prevent undefined input values
+    const defaultQuery: Partial<LokiQuery> = { expr: '', legendFormat: '' };
+    const query = Object.assign({}, defaultQuery, props.query);
+    this.query = query;
+    // Query target properties that are fully controlled inputs
+    this.state = {
+      // Fully controlled text inputs
+      legendFormat: query.legendFormat,
     };
+
+    const { data } = this.props;
+
+    if (data && data.request) {
```

Comment:
When the time range handling is moved into the constructor, the range won't be updated when someone changes the timepicker. This will affect the tab completion results.